internal/wasmldr/src/loader/compiled/tls.rs

```diff
@@ -0,0 +1,340 @@
+}
+
+impl<T> DerefMut for Forgotten<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.as_mut().unwrap()
+    }
+}
+
+impl<T> From<T> for Forgotten<T> {
+    fn from(value: T) -> Self {
+        Self(Some(value))
+    }
+}
+
+impl<T> Drop for Forgotten<T> {
+    fn drop(&mut self) {
+        forget(self.0.take());
+    }
+}
+
+pub struct Stream {
+    lck: RwLock<(Forgotten<CapStream>, Connection)>,
+    any: AnyStream,
+}
+
+impl From<Stream> for Box<dyn WasiFile> {
+    fn from(value: Stream) -> Self {
+        Box::new(value)
+    }
+}
+
+impl Stream {
+    fn new(tcp: CapStream, tls: Connection) -> Self {
+        let cap = unsafe { CapStream::from_raw_fd(tcp.as_raw_fd()) }.into();
+        let any = AnyStream::from_cap_std(tcp);
+        Self {
+            lck: RwLock::new((cap, tls)),
+            any,
+        }
+    }
+
+    pub fn connect(
+        mut tcp: cap_std::net::TcpStream,
+        name: &str,
+        cfg: Arc<ClientConfig>,
+    ) -> Result<Self, Error> {
+        // Set up connection.
+        let tls = ClientConnection::new(cfg, name.try_into()?)?;
+        let mut tls = Connection::Client(tls);
+
+        // Finish the connection.
+        tls.complete_io(&mut tcp)?;
+
+        Ok(Self::new(tcp, tls))
+    }
+}
+
+#[wiggle::async_trait]
+impl WasiFile for Stream {
+    fn as_any(&self) -> &dyn Any {
+        self.any.as_any()
+    }
+
+    async fn sock_accept(&mut self, fdflags: FdFlags) -> Result<Box<dyn WasiFile>, Error> {
+        self.any.sock_accept(fdflags).await
+    }
+
+    async fn datasync(&self) -> Result<(), Error> {
+        self.any.datasync().await
+    }
+
+    async fn sync(&self) -> Result<(), Error> {
+        self.any.sync().await
+    }
+
+    async fn get_filetype(&self) -> Result<FileType, Error> {
+        self.any.get_filetype().await
+    }
+
+    async fn get_fdflags(&self) -> Result<FdFlags, Error> {
+        self.any.get_fdflags().await
+    }
+
+    async fn set_fdflags(&mut self, fdflags: FdFlags) -> Result<(), Error> {
+        self.any.set_fdflags(fdflags).await
+    }
+
+    async fn get_filestat(&self) -> Result<Filestat, Error> {
+        self.any.get_filestat().await
+    }
+
+    async fn set_filestat_size(&self, size: u64) -> Result<(), Error> {
+        self.any.set_filestat_size(size).await
+    }
+
+    async fn advise(&self, offset: u64, len: u64, advice: Advice) -> Result<(), Error> {
+        self.any.advise(offset, len, advice).await
+    }
+
+    async fn allocate(&self, offset: u64, len: u64) -> Result<(), Error> {
+        self.any.allocate(offset, len).await
+    }
+
+    async fn set_times(
+        &self,
+        atime: Option<wasi_common::SystemTimeSpec>,
+        mtime: Option<wasi_common::SystemTimeSpec>,
+    ) -> Result<(), Error> {
+        self.any.set_times(atime, mtime).await
+    }
+
+    async fn read_vectored<'a>(&self, bufs: &mut [IoSliceMut<'a>]) -> Result<u64, Error> {
+        let (cap, tls) = &mut *self.lck.write().unwrap();
```

Comment:
This needs write access, not just read access?

Reply:
I think previous approach with `tcp` rather than `cap` was clearer